File: v1/LLD/Design-Patterns/Command/code/Command.cpp
// Command Pattern — a RemoteControl (Invoker) triggers Commands against a Light (Receiver), with undo support.
// Compile: g++ -std=c++14 Command.cpp -o command
// Run:     ./command

#include <iostream>
#include <vector>
#include <memory>

// Receiver — knows how to actually perform the operations.
class Light {
public:
    void turnOn() {
        on_ = true;
        std::cout << "Light is ON" << std::endl;
    }

    void turnOff() {
        on_ = false;
        std::cout << "Light is OFF" << std::endl;
    }

private:
    bool on_ = false;
};

class RemoteCommand {
public:
    virtual void execute() = 0;
    virtual void undo() = 0;
    virtual ~RemoteCommand() = default;
};

class LightOnCommand : public RemoteCommand {
public:
    explicit LightOnCommand(Light& light) : light_(light) {}

    void execute() override { light_.turnOn(); }
    void undo() override { light_.turnOff(); }

private:
    Light& light_;
};

class LightOffCommand : public RemoteCommand {
public:
    explicit LightOffCommand(Light& light) : light_(light) {}

    void execute() override { light_.turnOff(); }
    void undo() override { light_.turnOn(); }

private:
    Light& light_;
};

// Invoker — triggers commands and keeps a history for undo.
class RemoteControl {
public:
    void pressButton(RemoteCommand* command) {
        command->execute();
        history_.push_back(command);
    }

    void pressUndo() {
        if (history_.empty()) {
            std::cout << "Nothing to undo." << std::endl;
            return;
        }
        RemoteCommand* last = history_.back();
        history_.pop_back();
        last->undo();
    }

private:
    std::vector<RemoteCommand*> history_;
};

int main() {
    Light light;
    RemoteControl remote;

    LightOnCommand onCommand(light);
    LightOffCommand offCommand(light);

    std::cout << "Press ON button:" << std::endl;
    remote.pressButton(&onCommand);

    std::cout << "Press OFF button:" << std::endl;
    remote.pressButton(&offCommand);

    std::cout << "Press UNDO button:" << std::endl;
    remote.pressUndo();

    std::cout << "Press UNDO button again:" << std::endl;
    remote.pressUndo();

    std::cout << "Press UNDO button once more (nothing left):" << std::endl;
    remote.pressUndo();

    return 0;
}

File: v1/LLD/Design-Patterns/Command/code/command.js
/**
 * Command Pattern — a RemoteControl (Invoker) triggers Commands against a Light (Receiver), with undo support.
 * Run: node command.js
 */

class Light {
  constructor() {
    this.on = false;
  }

  turnOn() {
    this.on = true;
    console.log("Light is ON");
  }

  turnOff() {
    this.on = false;
    console.log("Light is OFF");
  }
}

class LightOnCommand {
  constructor(light) {
    this.light = light;
  }

  execute() {
    this.light.turnOn();
  }

  undo() {
    this.light.turnOff();
  }
}

class LightOffCommand {
  constructor(light) {
    this.light = light;
  }

  execute() {
    this.light.turnOff();
  }

  undo() {
    this.light.turnOn();
  }
}

class RemoteControl {
  constructor() {
    this.history = [];
  }

  pressButton(command) {
    command.execute();
    this.history.push(command);
  }

  pressUndo() {
    if (this.history.length === 0) {
      console.log("Nothing to undo.");
      return;
    }
    const last = this.history.pop();
    last.undo();
  }
}

const light = new Light();
const remote = new RemoteControl();

const onCommand = new LightOnCommand(light);
const offCommand = new LightOffCommand(light);

console.log("Press ON button:");
remote.pressButton(onCommand);

console.log("Press OFF button:");
remote.pressButton(offCommand);

console.log("Press UNDO button:");
remote.pressUndo();

console.log("Press UNDO button again:");
remote.pressUndo();

console.log("Press UNDO button once more (nothing left):");
remote.pressUndo();

module.exports = { Light, LightOnCommand, LightOffCommand, RemoteControl };
